File: prefix_suffix_algorithms/practice/product_except_self.py
"""
Problem: Product of Array Except Self (LeetCode 238)

Given an integer array nums, return an array answer such that answer[i] is equal 
to the product of all the elements of nums except nums[i].

The product of any prefix or suffix of nums is guaranteed to fit in a 32-bit integer.
You must write an algorithm that runs in O(n) time and without using the division operation.

Concept:
answer[i] = (Product of all left elements) * (Product of all right elements)
We can use a prefix product array and a suffix product array.
To optimize space to O(1) (excluding output array), we can calculate left products 
directly into the result array, and then multiply by right products on the fly.

Time Complexity: O(N)
Space Complexity: O(1) extra space (result array doesn't count).
"""

from typing import List

def product_except_self(nums: List[int]) -> List[int]:
    n = len(nums)
    answer = [1] * n
    
    # Pass 1: Calculate left products (prefix products)
    # answer[i] will contain product of nums[0...i-1]
    left_product = 1
    for i in range(n):
        answer[i] = left_product
        left_product *= nums[i]
        
    # Pass 2: Calculate right products (suffix products) and multiply
    # We maintain a running right_product
    right_product = 1
    for i in range(n - 1, -1, -1):
        answer[i] *= right_product
        right_product *= nums[i]
        
    return answer

# Test cases
if __name__ == "__main__":
    test1 = [1, 2, 3, 4]
    print(f"Arr: {test1} -> {product_except_self(test1)}") # Expected: [24, 12, 8, 6]
    
    test2 = [-1, 1, 0, -3, 3]
    print(f"Arr: {test2} -> {product_except_self(test2)}") # Expected: [0, 0, 9, 0, 0]

File: prefix_suffix_algorithms/practice/product_except_self.js
/**
 * Problem Link: https://leetcode.com/problems/product-of-array-except-self/
 * Difficulty: Medium
 * 
 * Problem Description:
 * Return an array where each `answer[i]` is the product of all elements except `nums[i]`.
 * Must be O(N) time and O(1) extra space (no division allowed).
 * 
 * Key Concepts:
 * - Prefix Product and Suffix Product
 * - In-place Computation
 * 
 * Intuition:
 * `Result[i] = (Left Product) * (Right Product)`
 * Left Product at i = Product of `0` to `i-1`.
 * Right Product at i = Product of `i+1` to `N-1`.
 * 
 * We can compute all Left Products first and store them in the result array.
 * Then, we can iterate backwards, keeping a running "Right Product" and multiplying it into the result array.
 * 
 * Algorithm:
 * 1. Fill `answer` with 1s.
 * 2. Iterate `i` from 0 to N: `answer[i] = leftProduct`. `leftProduct *= nums[i]`.
 * 3. Iterate `i` from N-1 down to 0: `answer[i] *= rightProduct`. `rightProduct *= nums[i]`.
 * 
 * Time Complexity: O(N) - Two passes.
 * Space Complexity: O(1) - Ignoring output array.
 */

/**
 * @param {number[]} nums
 * @return {number[]}
 */
function productExceptSelf(nums) {
    const n = nums.length;
    // Initialize answer array
    const answer = new Array(n).fill(1);

    // Pass 1: Calculate Left Products
    let leftProduct = 1;
    for (let i = 0; i < n; i++) {
        answer[i] = leftProduct;
        leftProduct *= nums[i];
    }

    // Pass 2: Calculate Right Products & Multiply
    let rightProduct = 1;
    for (let i = n - 1; i >= 0; i--) {
        // Multiply by the product of all elements to the right
        answer[i] *= rightProduct;

        // Accumulate current number for the next position
        rightProduct *= nums[i];
    }

    return answer;
}

// ---------------------------------------------------
// Test Cases
// ---------------------------------------------------
const test1 = [1, 2, 3, 4];
console.log(`Test 1: [${test1}] -> ${productExceptSelf(test1)}`); // Expected: [24, 12, 8, 6]

const test2 = [-1, 1, 0, -3, 3];
console.log(`Test 2: [${test2}] -> ${productExceptSelf(test2)}`); // Expected: [0, 0, 9, 0, 0]
